File: apps/Pose-Motion/iOS/PoseMotion/Overlay/Skeleton3DCanvas.swift
import SwiftUI

/// Side-view 3D skeleton: orthographic projection of the root-relative H36M joints,
/// rotated about the vertical axis by a drag-controlled azimuth.
struct Skeleton3DCanvas: View {
    let joints: [SIMD3<Float>]        // 17 H36M joints, root-relative
    @State private var azimuth: Double = 0.9   // radians; start slightly rotated
    @State private var dragStartAzimuth: Double?

    private static let bones: [(Int, Int)] = [
        (0, 1), (1, 2), (2, 3),          // right leg
        (0, 4), (4, 5), (5, 6),          // left leg
        (0, 7), (7, 8), (8, 9), (9, 10), // spine → head
        (8, 11), (11, 12), (12, 13),     // left arm
        (8, 14), (14, 15), (15, 16),     // right arm
    ]
    private static let leftJoints: Set<Int> = [4, 5, 6, 11, 12, 13]
    private static let rightJoints: Set<Int> = [1, 2, 3, 14, 15, 16]

    var body: some View {
        Canvas { context, size in
            guard joints.count == 17 else { return }
            let cosA = Float(cos(azimuth))
            let sinA = Float(sin(azimuth))

            // Rotate about Y (vertical); keep image-style y-down.
            let rotated: [(x: Float, y: Float, z: Float)] = joints.map {
                (x: $0.x * cosA + $0.z * sinA, y: $0.y, z: -$0.x * sinA + $0.z * cosA)
            }

            // Fit to canvas.
            let maxExtent = rotated.reduce(Float(1e-4)) { m, p in
                max(m, abs(p.x), abs(p.y))
            }
            let scale = CGFloat(0.42) * min(size.width, size.height) / CGFloat(maxExtent)
            let center = CGPoint(x: size.width / 2, y: size.height / 2)
            func project(_ p: (x: Float, y: Float, z: Float)) -> CGPoint {
                CGPoint(x: center.x + CGFloat(p.x) * scale, y: center.y + CGFloat(p.y) * scale)
            }

            // Depth-sorted bones, nearer drawn brighter and on top.
            let sortedBones = Self.bones.sorted {
                (rotated[$0.0].z + rotated[$0.1].z) > (rotated[$1.0].z + rotated[$1.1].z)
            }
            let zRange = rotated.map(\.z)
            let zMin = zRange.min() ?? 0
            let zMax = zRange.max() ?? 1
            func depthAlpha(_ z: Float) -> Double {
                let t = (z - zMin) / max(1e-4, zMax - zMin)
                return 0.35 + 0.65 * Double(1 - t)
            }

            for (a, b) in sortedBones {
                var path = Path()
                path.move(to: project(rotated[a]))
                path.addLine(to: project(rotated[b]))
                let color: Color =
                    Self.leftJoints.contains(b) ? Theme.leftSide :
                    Self.rightJoints.contains(b) ? Theme.rightSide : Theme.torso
                let alpha = depthAlpha((rotated[a].z + rotated[b].z) / 2)
                context.stroke(path, with: .color(color.opacity(alpha)),
                               style: StrokeStyle(lineWidth: 3, lineCap: .round))
            }
            for p in rotated {
                let pt = project(p)
                context.fill(
                    Path(ellipseIn: CGRect(x: pt.x - 2.5, y: pt.y - 2.5, width: 5, height: 5)),
                    with: .color(Theme.textPrimary.opacity(depthAlpha(p.z)))
                )
            }
        }
        .gesture(
            DragGesture(minimumDistance: 2)
                .onChanged { value in
                    if dragStartAzimuth == nil { dragStartAzimuth = azimuth }
                    azimuth = (dragStartAzimuth ?? azimuth) + Double(value.translation.width) * 0.012
                }
                .onEnded { _ in dragStartAzimuth = nil }
        )
    }
}

File: apps/Pose-Motion/Android/app/src/main/java/ai/zetic/demo/posemotion/ui/overlay/Skeleton3dCanvas.kt
package ai.zetic.demo.posemotion.ui.overlay

import ai.zetic.demo.posemotion.pipeline.Vec3
import ai.zetic.demo.posemotion.ui.Theme
import androidx.compose.foundation.Canvas
import androidx.compose.foundation.gestures.detectDragGestures
import androidx.compose.runtime.Composable
import androidx.compose.runtime.getValue
import androidx.compose.runtime.mutableFloatStateOf
import androidx.compose.runtime.remember
import androidx.compose.runtime.setValue
import androidx.compose.ui.Modifier
import androidx.compose.ui.geometry.Offset
import androidx.compose.ui.graphics.Color
import androidx.compose.ui.graphics.StrokeCap
import androidx.compose.ui.input.pointer.pointerInput
import kotlin.math.abs
import kotlin.math.cos
import kotlin.math.sin

/**
 * Side-view 3D skeleton: orthographic projection of the root-relative H36M joints,
 * rotated about the vertical axis by a drag-controlled azimuth.
 */
@Composable
fun Skeleton3dCanvas(joints: List<Vec3>, modifier: Modifier = Modifier) {
    var azimuth by remember { mutableFloatStateOf(0.9f) }   // radians; start slightly rotated

    Canvas(
        modifier = modifier.pointerInput(Unit) {
            detectDragGestures { change, dragAmount ->
                change.consume()
                azimuth += dragAmount.x * 0.012f
            }
        }
    ) {
        if (joints.size != 17) return@Canvas
        val cosA = cos(azimuth)
        val sinA = sin(azimuth)

        // Rotate about Y (vertical); keep image-style y-down.
        val rotated = joints.map {
            Vec3(it.x * cosA + it.z * sinA, it.y, -it.x * sinA + it.z * cosA)
        }

        val maxExtent = rotated.fold(1e-4f) { m, p -> maxOf(m, abs(p.x), abs(p.y)) }
        val scale = 0.42f * minOf(size.width, size.height) / maxExtent
        val center = Offset(size.width / 2, size.height / 2)
        fun project(p: Vec3) = Offset(center.x + p.x * scale, center.y + p.y * scale)

        val zMin = rotated.minOf { it.z }
        val zMax = rotated.maxOf { it.z }
        fun depthAlpha(z: Float): Float {
            val t = (z - zMin) / maxOf(1e-4f, zMax - zMin)
            return 0.35f + 0.65f * (1f - t)
        }

        // Depth-sorted bones, nearer drawn brighter and on top.
        val sortedBones = BONES.sortedByDescending { rotated[it.first].z + rotated[it.second].z }
        for ((a, b) in sortedBones) {
            val color: Color = when {
                b in LEFT_JOINTS -> Theme.LeftSide
                b in RIGHT_JOINTS -> Theme.RightSide
                else -> Theme.Torso
            }
            val alpha = depthAlpha((rotated[a].z + rotated[b].z) / 2)
            drawLine(
                color.copy(alpha = alpha),
                project(rotated[a]), project(rotated[b]),
                strokeWidth = 3f, cap = StrokeCap.Round,
            )
        }
        for (p in rotated) {
            drawCircle(Theme.TextPrimary.copy(alpha = depthAlpha(p.z)), 2.5f, project(p))
        }
    }
}

private val BONES = listOf(
    0 to 1, 1 to 2, 2 to 3,           // right leg
    0 to 4, 4 to 5, 5 to 6,           // left leg
    0 to 7, 7 to 8, 8 to 9, 9 to 10,  // spine → head
    8 to 11, 11 to 12, 12 to 13,      // left arm
    8 to 14, 14 to 15, 15 to 16,      // right arm
)
private val LEFT_JOINTS = setOf(4, 5, 6, 11, 12, 13)
private val RIGHT_JOINTS = setOf(1, 2, 3, 14, 15, 16)
